src/erc7496/erc7496.cairo

```diff
@@ -0,0 +1,155 @@
+            self.ERC7498_trait_metadata_uri.read()
+        }
+    }
+
+    //
+    // Internal
+    //
+
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl SRC5: SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Initializes the contract
+        /// This should only be used inside the contract's constructor.
+        fn initializer(ref self: ComponentState<TContractState>) {
+            let mut src5_component = get_dep_component_mut!(ref self, SRC5);
+            src5_component.register_interface(IERC7496_ID);
+        }
+
+        // @notice Get the value of a trait for a given token ID.
+        // @param tokenId The token ID to get the trait value for
+        // @param traitKey The trait key to get the value of
+        fn _get_trait_value(
+            self: @ComponentState<TContractState>, token_id: u256, trait_key: felt252
+        ) -> felt252 {
+            // Return the trait value.
+            self.ERC7496_traits.read((token_id, trait_key))
+        }
+
+        // @notice Get the values of traits for a given token ID.
+        // @param tokenId The token ID to get the trait values for
+        // @param traitKeys The trait keys to get the values of
+        fn _get_trait_values(
+            self: @ComponentState<TContractState>, token_id: u256, trait_keys: Span<felt252>
+        ) -> Span<felt252> {
+            let mut trait_values = array![];
+            // Assign each trait value to the corresponding key.
+            let mut index = 0;
+            while index < trait_keys
+                .len() {
+                    trait_values
+                        .append(self.ERC7496_traits.read((token_id, *trait_keys.at(index))));
+                    index += 1;
+                };
+            trait_values.span()
+        }
+
+        // @notice Set the value of a trait for a given token ID.
+        //         Reverts if the trait value is unchanged.
+        // @dev    IMPORTANT: Override this method with access role restriction.
+        // @param tokenId The token ID to set the trait value for
+        // @param traitKey The trait key to set the value of
+        // @param newValue The new trait value to set
+        fn _set_trait(
+            ref self: ComponentState<TContractState>,
+            token_id: u256,
+            trait_key: felt252,
+            new_value: felt252
+        ) {
+            // Revert if the new value is the same as the existing value.
+            let existing_value = self.ERC7496_traits.read((token_id, trait_key));
+            assert(existing_value != new_value, Errors::TRAIT_VALUE_UNCHANGED);
+            // Set the new trait value.
+            self._set_trait_internal(token_id, trait_key, new_value);
+            // Emit the event noting the update.
+            self.emit(TraitUpdated { trait_key, token_id, trait_value: new_value });
```

Comment:
mmh i think the event should be emitted in the public `set_trait` function in the preset. 
This might lead to double event emission (if i want to implement a contract  using this component, with a `set_traits` function that loops on a range to modify multiple traits and emit the `TraitUpdatedRange` event of the IERC7496, then i will also emit the `TraitUpdated` for each iteration which is superfluous